client/web/src/components/widgets/options/ToolOptions.vue

```diff
@@ -0,0 +1,131 @@
+
+interface IconOption {
+	kind: "icon";
+	icon: string;
+	title: string;
+}
+
+interface SeparatorOption {
+	kind: "separator";
+	type: SeparatorType;
+}
+
+interface PopoverOption {
+	kind: "popover";
+	title: string;
+	placeholder_text: string;
+}
+
+interface NumberOption {
+	kind: "number";
+	initial: number;
+	step: number;
+	min?: number;
+	callback?: Function;
+}
+
+export default defineComponent({
+	props: {
+		activeTool: { type: String },
+	},
+	computed: {},
+	methods: {
+		async setToolOptions(new_value: number) {
+			// TODO: Each value-input widget (i.e. not a button) should map to a field in an options struct,
+			// and updating a widget should send the whole updated struct to the backend.
+			// Later, it could send a single-field update to the backend.
+			const { set_tool_options } = await wasm;
+			set_tool_options(this.$props.activeTool || "Select", { Shape: { shape_type: { Polygon: { vertices: new_value } } } });
+		},
+	},
+	data() {
+		return {
+			optionsMap: new Map([
+				[
+					"Select",
+					[
+						{ kind: "icon", icon: "AlignHorizontalLeft", title: "Horizontal Align Left" },
+						{ kind: "icon", icon: "AlignHorizontalCenter", title: "Horizontal Align Center" },
+						{ kind: "icon", icon: "AlignHorizontalRight", title: "Horizontal Align Right" },
+
+						{ kind: "separator", type: SeparatorType.Unrelated },
+
+						{ kind: "icon", icon: "AlignVerticalTop", title: "Vertical Align Top" },
+						{ kind: "icon", icon: "AlignVerticalCenter", title: "Vertical Align Center" },
+						{ kind: "icon", icon: "AlignVerticalBottom", title: "Vertical Align Bottom" },
+
+						{ kind: "separator", type: SeparatorType.Related },
+
+						{ kind: "popover", title: "Align", placeholder_text: "More alignment-related buttons will be here" },
+
+						{ kind: "separator", type: SeparatorType.Section },
+
+						{ kind: "icon", icon: "FlipHorizontal", title: "Flip Horizontal" },
+						{ kind: "icon", icon: "FlipVertical", title: "Flip Vertical" },
+
+						{ kind: "separator", type: SeparatorType.Related },
+
+						{ kind: "popover", title: "Flip", placeholder_text: "More flip-related buttons will be here" },
+
+						{ kind: "separator", type: SeparatorType.Section },
+
+						{ kind: "icon", icon: "BooleanUnion", title: "Boolean Union" },
+						{ kind: "icon", icon: "BooleanSubtractFront", title: "Boolean Subtract Front" },
+						{ kind: "icon", icon: "BooleanSubtractBack", title: "Boolean Subtract Back" },
+						{ kind: "icon", icon: "BooleanIntersect", title: "Boolean Intersect" },
+						{ kind: "icon", icon: "BooleanDifference", title: "Boolean Difference" },
+
+						{ kind: "separator", type: SeparatorType.Related },
+
+						{ kind: "popover", title: "Boolean", placeholder_text: "More boolean-related buttons will be here" },
+					],
+				],
+				["Shape", [{ kind: "number", initial: 6, step: 1, min: 3, callback: this.setToolOptions }]],
+			]) as ToolOptionsMap,
```

Comment:
I investigated the same thing when I was writing this, but I couldn't find a solution that does type checking. Let me know if you have any ideas to refactor it, otherwise this will be fixed as we migrate widget specification to Rust.

Reply:
Typically in TS it's a matter of changing the fist snippet to the second snippet:

```ts
return fooBar as ToolOptionsMap;
```

```ts
const x: ToolOptionsMap = fooBar;
return x;
```

I haven't looked closely enough to see if that doesn't apply in this case though.

Reply:
That works beautifully, thanks